File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/projection/ProjectionOpExec.java
package edu.uci.ics.texera.workflow.operators.projection;

import edu.uci.ics.texera.workflow.common.operators.map.MapOpExec;
import edu.uci.ics.texera.workflow.common.tuple.Tuple;
import scala.Function1;
import scala.Serializable;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class ProjectionOpExec extends MapOpExec {
    private final ProjectionOpDesc opDesc;

    public ProjectionOpExec(ProjectionOpDesc opDesc) {
        this.opDesc = opDesc;
        this.setMapFunc((Function1<Tuple, Tuple> & Serializable) this::processTuple);
    }

    public Tuple processTuple(Tuple t) {
        List<String> attributes = opDesc.attributes;
        List<String> attributesToRemove = t.getSchema().getAttributeNames().stream()
                .filter(item -> !attributes.contains(item))
                .collect(Collectors.toList());

        return Tuple.newBuilder().add(t).removeIfExists(attributesToRemove).build();
    }

}

File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/projection/ProjectionOpExec.scala
package edu.uci.ics.texera.workflow.operators.projection

import com.google.common.base.Preconditions
import edu.uci.ics.amber.engine.common.model.tuple.{Tuple, TupleLike}
import edu.uci.ics.texera.workflow.common.operators.map.MapOpExec

import scala.collection.mutable

class ProjectionOpExec(
    attributeUnits: List[AttributeUnit],
    dropOption: Boolean = false
) extends MapOpExec {

  setMapFunc(project)
  def project(tuple: Tuple): TupleLike = {
    Preconditions.checkArgument(attributeUnits.nonEmpty)
    var selectedUnits: List[AttributeUnit] = List()
    val fields = mutable.LinkedHashMap[String, Any]()
    if (dropOption) {
      val allAttribute = tuple.schema.getAttributeNames
      val selectedAttributes = attributeUnits.map(_.getOriginalAttribute)
      val keepAttributes = allAttribute.diff(selectedAttributes)

      keepAttributes.foreach { attribute =>
        val newList = List(
          new AttributeUnit(attribute, attribute)
        )
        selectedUnits = selectedUnits ::: newList
      }

    } else {

      selectedUnits = attributeUnits
    }

    selectedUnits.foreach { attributeUnit =>
      val alias = attributeUnit.getAlias
      if (fields.contains(alias)) {
        throw new RuntimeException("have duplicated attribute name/alias")
      }
      fields(alias) = tuple.getField[Any](attributeUnit.getOriginalAttribute)
    }

    TupleLike(fields.toSeq: _*)
  }

}
